File: my-code/6/main.py
# -*- coding:utf-8 -*-
# @Time     : 2021/05/18 18: 56
# @Author   : Ranshi
# @File     : 6. Z 字形变换.py
class Solution:
    def convert(self, s: str, num_rows: int) -> str:
        if num_rows == 1:
            return s
        path = [[] for _ in range(num_rows)]
        idx, mode = 0, 1
        for item in s:
            path[idx].append(item)
            if mode == 1 and idx == num_rows - 1:
                mode = 2
            elif mode == 2 and idx == 0:
                mode = 1
            if mode == 1:
                idx += 1
            else:
                idx -= 1
        return "".join("".join(item) for item in path)


if __name__ == "__main__":
    print(Solution().convert(s="PAYPALISHIRING", num_rows=3))

File: my-code/6/main.js
// @Time     : 2021/05/28 17: 47
// @Author   : Ranshi
// @File     : 6. Z 字形变换.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
const convert = function (s, numRows) {
  if (numRows === 1) {
    return s;
  }
  const len = Math.min(s.length, numRows),
    rows = [];
  for (let i = 0; i < len; i++) {
    rows[i] = '';
  }
  let loc = 0,
    down = false;
  for (const c of s) {
    rows[loc] += c;
    if (loc === 0 || loc === numRows - 1) {
      down = !down;
    }
    loc += down ? 1 : -1;
  }
  let ans = '';
  for (const row of rows) {
    ans += row;
  }
  return ans;
};

console.log(convert((s = 'PAYPALISHIRING'), (numRows = 4)));
